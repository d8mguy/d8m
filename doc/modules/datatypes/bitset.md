## Bitset

Module bitset defines the `bitset0` and `smallint64` types. `Bitset0` provides bitsets in single 64-bit integer words; `smallint64` defines integers taking values between 0 and 63.

The compiler interacts indirectly with these definitions by converting ELTs (enumerated literal types) internally to integer and asserting of them the max values they can take. So ELTs with fewer than 64 elements will work as bitset0. The module defines a STMapping for `set(smallint64)` and ensures that it's defined whenever type `set` is available. However, to use it you must `applyST` yourself.

You can enumerate a `bitset0`; the definition uses the eachstart protocol.

    [bitset0: lst]

where `lst: list(integer)`.

    val bitset0 = extend tuple(...private...) where {
        method setbit = \mod(inx: integer) -> nothing {...}
        method clearbit = \mod(inx: integer) -> nothing {...}
        method empty = \() -> boolean {...}
        method inBitSet = \(inx: integer) -> boolean {...}
        method count = \() -> integer {...}
        method union = \(other: bitset0) -> bitset0 {...}
        method intersection = \(other: bitset0) -> bitset0 {...}
        method subset = \(other: bitset0) -> boolean {...}
    }

Method `subset` tells if self is a subset of other. All the operations here are O(1); count uses the `OnesCount` function in go's `bits` package which in turn uses a hardware instruction if available, so it should be O(1) as well. 