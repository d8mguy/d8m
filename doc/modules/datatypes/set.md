## Set

Module set provides an abstract definition of set and an STMap to go's map type.

    [set(T): lst]

where `lst:list(T)`.

    val set = \typegen(T::entity) {
        val setT = extend tuple(...private...) where {
            method $cvt = \() -> list(T) {...}
            method $in = \(elt: T) -> boolean {...}
            method add = \mod(elt: T) -> nothing {...}
            method add = \mod(other: setT) -> nothing {...}
            method remove = \mod(elt: T) -> nothing {...}
            method remove = \mod(elts: list(T)) -> nothing {...}
            method count = \() -> integer {...}
            method isEmpty = \() -> boolean {...}
            method overlaps = \imp(other: setT) -> boolean {...}
            method $= = \(other: setT) -> boolean {...}
            method subset = \(other: setT) -> boolean {...}
            method intersection = \(other: setT) -> setT {...}
            method $+ = \(other: T) -> setT {...}
            method $+ = \(other: setT) -> setT {...}
        }
    }
