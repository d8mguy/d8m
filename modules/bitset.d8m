// Module bitset defines the bitset0 and smallint64 types. Bitset0 provides bitsets in single integer64 words; smallint64
// defines integers taking values between 0 and 63.
// The compiler interacts indirectly with these definitions by converting ELTs internally to integer and asserting of them the max
// values they can take. So ELTs with fewer than 64 elements will work as bitset0. The module defines a STMapping for
// set(smallint64) and ensures that it's processed whenever the set type is present. However, to use it you must applyST yourself.
//
// Bitset0 uses eachstart protocol to be enumerable.

import go "math/bits"
val smallint64 = extend integer where { assert self >= 0 && self < 64 }
val bitset0 =
  extend tuple(private x: bits64) where {
    method setbit = \mod(inx: integer) { assert inx < 64; x = or(x, bitnum(inx)) }
    method clearbit = \mod(inx: integer) { assert inx < 64; x = andnot(x, bitnum(inx)) }
    method empty = \() { cvt(x, integer) == 0 }
    method inBitSet = \(inx: integer) { cvt(and(x, bitnum(inx)), integer) != 0 }
    method count = \() { bits.OnesCount(cvt(x, integer)) }
    method union = \imp(other: bitset0) { var rslt = copy(self); rslt.x = or(x, other.x); rslt }

    // tells if self is a subset of other
    method contains = \(other: bitset0) { and(x, other.x) == x }
    method intersection = \imp(other: bitset0) { var rslt = copy(self); rslt.x = and(rslt.x, other.x); rslt }
    method lithook = \mod(lst: list(integer)) { each(elt^lst) setbit(elt); self }
    method eachstart = \imp() {
        val bs = self
        val estype = extend tuple(nelts, seen, nxinx: integer, done: boolean) where {
            method value = \() { nxinx }
            method eachdone = \() { done }
            method eachstep = \mod() {
                seen += 1
                if(seen == nelts) done = true
                else loop { nxinx += 1; if(bs.inBitSet(nxinx)) break }
            }
            method lithook = \mod(nelts: integer) {
                self.done = nelts <= 0
                self.nelts = nelts
                self.seen = 0
                self.nxinx = 0
                unless(done) { each(i^0..64) if(bs.inBitSet(i)) { nxinx = i; break } }
            }
        }
        return [estype: count(self)]
    }
  }

oncondition defined(set, type) {
    given(aBit: integer, aSet: set(smallint64)) {
    	STMap {
    		source = set(smallint64)
    		target = bitset0
    		direct count, lithook
    		aBit in self => self.inBitSet(aBit)
    		self.isEmpty => self.empty
    		self.add(aBit) => self.setbit(aBit)
    		self.remove(aBit) => self.clearbit(aBit)
    		// the cast in rhs ensures that both operands to union are bitsets
    		self + aSet => union(self, aSet)
    		intersection(self, aSet) => intersection(self, aSet)
    		subset(self, aSet) => contains(self, aSet)
    		enumerable(self) => self
    	}
    }
}
export bitset0, smallint64

/*
import "testing" melted
val tests = \mod(options: list(label)) {
	var basics = [tstate: :basics]
	if(:basics in options) {
		var bs1 = [bitset0: 1,5,2]
		var bs2 = [bitset0: 7,3,19,21,0]
		var bs3 = [bitset0:]
		basics.should(bs3.empty, "true")
		basics.should(bs1.empty, "false")
		basics.should(bs1.union(bs3), "38")
		basics.should(bs2.union(bs3), "2621577")
		basics.should(bs2.union(bs1), "2621615")
		basics.should(bs2.intersection(bs3).empty, "true")
		bs2.clearbit(21)
		basics.should(bs2.union(bs3), "524425")
		bs2.setbit(3)
		basics.should(bs2.union(bs3), "524425")
		basics.should(bs2.inBitSet(1), "false")
		basics.should(bs2.inBitSet(2), "false")
		basics.should(bs2.inBitSet(3), "true")
		basics.should(bs3.inBitSet(14), "false")
		basics.should(bs1.inBitSet(6), "false")
		basics.should(bs1.inBitSet(5), "true")
		bs2.clearbit(4)
		basics.should(bs2.union(bs3), "524425")
		bs2.clearbit(19)
		basics.should(bs2.union(bs3), "137")
		basics.should(bs1.union(bs2), "175")
		bs2.setbit(5)
		basics.should(bs2.union(bs3), "169")
		basics.should(bs1.union(bs2), "175")	
	}
	basics.report
}
*/

