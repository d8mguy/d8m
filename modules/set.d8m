// Module set provides an abstract definition of set and an STMap to go's map type.

export set

val set = \typegen(T::entity) {
	val setT = extend tuple(private x: list(T)) where {
		assert enumerable(x)
		method $cvt = \() -> list(T) { copy(x) }
		method $in = \(elt: T) { elt in x }
		method add = \mod(elt: T) { unless(elt in x) x.pushb(elt) }
		method add = \mod(other: setT) { each(o^other) self.add(o) }
		method remove = \mod(elt: T) { val inx = x.index(elt); if(inx != nil) x.removeIndex(inx) }
		method remove = \mod(elts: list(T)) { each(e^elts) self.remove(e) }
		method lithook = \mod(c: list(T)) { each(x^c) add(x) }
		method count = \() { x.count }
		method isEmpty = \() { x.count == 0 }
		method overlaps = \imp(other: setT) { var rslt = false; each(e^x) if(e in other) { rslt = true; break }; rslt }
		method $== = \(other: setT) -> boolean { other.count == x.count && all(self.{this in other}) }
		// Note: this is self <= other.
		method subset = \(other: setT) { all(self.{this in other}) }
		// Note: intersection could be more efficient, no need for the uniq-check in "rslt.add(o)". But there are tradeoffs.
		method intersection = \imp(other: setT) { var rslt: setT = []; each(o^other) if(o in x) rslt.add(o); rslt }
		method $+ = \imp(other: T) { var rslt: setT = copy(self); rslt.add(other); rslt }
		method $+ = \imp(other: setT) { var rslt: setT = copy(self); each(o^other) rslt.add(o); rslt }
	}
	setT
}
given(T1::entity, y: T1, aLT: list(T1)) {
	STMap {
		source = set(T1)
		target = Gomap(T1, boolean)
		direct count
		y in self => rvalindexWD(self, y, false)
		self.add(y) => self.lvalindex(y, true)
		self.remove(y) => self.delete(y)
		enumerable(self) => self.keys
	}
}
