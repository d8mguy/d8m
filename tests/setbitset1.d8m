import "bitset" melted
import "set" melted
val quintiles = label(:first, :second, :third, :fourth, :fifth)
val qset = set(quintiles)
applyST(bitset0, \(s: qset) { true } )
val s0 = [qset: :first, :third]
val tryQset = \() {
    val oq = [qset: :second, :first, :fifth]
    println("first subset: ", subset(s0, oq))
    println("intersect: ", intersection(s0, oq))
}
tryQset()