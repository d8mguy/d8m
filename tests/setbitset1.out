first subset:  false
intersect:  [1]
